### PR TITLE
daemon, snappy: add find (aka search)

### PR DESCRIPTION
**Pull request by chipaca**
Give snapd the ability to search by package name.

### CODE REVIEW

**Comment by mvo5** [comment]
Looks good

File: daemon/api.go [hunk 1]
```diff
@@ -303,16 +305,32 @@ func getSnapsInfo(c *Command, r *http.Request) Response {
...
-		// TODO: If there are no results (local or remote), report the error. If
-		//       there are results at all, inform that the result is partial.
-		found, _ := newRemoteRepo().All()
-		if len(found) > 0 {
+		repo := newRemoteRepo()
+		var found []snappy.Part
+
+		if searchTerm != "" {
+			found, _ = repo.Find(searchTerm)
 			sources = append(sources, "store")
```
Comment (pedronis) [inline daemon/api.go]: why is this one not conditional on the size of found? like below, or the one below unconditional, or should it be conditional only if sources wasn't given to start with and we used the fallback

**Comment by pedronis** [comment]
looks good to me but a wondering about the setting of sources to report back

File: daemon/api.go (hunk 1 above)
Comment (chipaca) [inline daemon/api.go]: Yes, I've been thinking about this. I think we want sources to indicate
whether we successfully queried the source (whether it returned anything it
not), and a new warnings field that lists things life the store being
unreachable. Perhaps also a per snap sources to say where each snap info
comes from

Comment (pedronis) [inline daemon/api.go]: @chipaca seems reasonable, so making it unconditional in terms of hits and then a TODO abou that other stuff?

Comment (chipaca) [inline daemon/api.go]: Yup

File: daemon/api.go [hunk 2]
```diff
@@ -173,11 +173,11 @@ func sysInfo(c *Command, r *http.Request) Response {
 type metarepo interface {
 	Details(string, string) ([]snappy.Part, error)
 	All() ([]snappy.Part, error)
-	Updates() ([]snappy.Part, error)
+	Find(string) ([]snappy.Part, error) // XXX: change this to Search iff aliases (and thus SharedNames) are no more
```
Comment (niemeyer) [inline daemon/api.go]: Can we have a slightly less enigmatic comment here? :-)

The note was useful for me to go digging further, but it'd be nice to be more clear here. Is that about the fact we have a Search method already? If so, can we have slightly more clear names for the time being, which highlight the distinction between Find and Search?

File: daemon/api.go [hunk 3]
```diff
@@ -303,16 +305,32 @@ func getSnapsInfo(c *Command, r *http.Request) Response {
...
+		repo := newRemoteRepo()
+		var found []snappy.Part
+
+		if searchTerm != "" {
+			found, _ = repo.Find(searchTerm)
 			sources = append(sources, "store")
+		} else {
+			found, _ = repo.All()
+
+			// TODO: If there are no results (local or remote), report the error. If
```
Comment (niemeyer) [inline daemon/api.go]: This TODO was about the error ignored above. Can we please hold it above the offending line so we spot what this is about more easily?

Comment (niemeyer) [inline daemon/api.go]: Sorry, I missed the fact we're _also_ querying it for All just here. Why are we doing that?  Can't we `Find` at all times, even if the searchTerm is empty?  Is there a reason not to do that?

Alternatively, the TODO above is now about both `Find` and `All`, so would be good to have a more clear comment indicating the problem right after includeStore conditional.

**Comment by niemeyer** [comment]
Looks nice. Wondering about one detail, per above note.

**Pull request merged**